rfcs/react-components/convergence/custom-styling.md

```diff
@@ -0,0 +1,510 @@
+context, and context provider.
+
+```tsx
+export type ComponentCustomStylesContextValue = {
+  useCustomStyles?: (state: ComponentState) => ComponentState;
+};
+
+export const ComponentCustomStylesContext: Context<ComponentCustomStylesContextValue> = createContext<ComponentCustomStylesContextValue>(
+  {},
+);
+
+export const ComponentCustomStylesContextProvider = ComponentCustomStylesContext.Provider;
+```
+
+When a caller wanted to customize the component, they use the provider. Callers would be able to control custom
+styling scope just like Fluent Provider. While they would need to use a provider per custom styled component,
+these could be aggregated into a component that transcludes children.
+
+```tsx
+<App>
+  <FluentProvider theme={webLightTheme}>
+    <ComponentCustomStylesContextProvider value={{ useCustomStyles }}>
+      {/* other component hierarchy */}
+      <Component>Hello custom styles!</Component>
+    </ComponentCustomStylesContextProvider>
+  </FluentProvider>
+</App>
+```
+
+### 👍
+
+- Custom styling gets the same power as the default component useStyles hook.
+- This builds on using slots as the addressable items for applying styles.
+- The custom styles can be built using Griffel with build-time optimization.
+- Can be applied at different scope like FluentProvider.
+- Does not require composing or introducing a new component type that must be adopted in lieu of the default component.
+- Fewer CSS specificity problems as styles can get overwritten.
+- Provides the type-safety expected from Fluent UI React.
+- Minimal code to add to each component. Could be put into the code generation template.
+- Does not significantly affect rendering performance.
+- Partners building components can follow the same practice to support custom styling in their libraries.
+
+### 👎
+
+- Custom styling all components requires adding and configuring lots of context providers.
+- Requires writing new code for every component, adding a story, and documenting the feature.
+- Context is harder to discover as it is outside the props contract.
+- Callers doing custom styling may still take a dependency on the CSS internals of the component if they don't replace a
+  style block completely.
+
+### 🤔 Centralized and singular custom styling method
+
+It would be nice to have a single method on context that could apply for all components. This proves difficult
+with each component having its own props and state types.
+
+Another approach would be to define single context value type in the react-shared-contexts package that had
+a method for each component, but that causes a circular dependency problem for components.
+
+There may some typescript trickery that would allow for a single context value type, but having each component define
+the context type it will inspect for custom styling follows SOLID dependency inversion.
+
+## Option A2: Add global custom styling hook
```

Comment:
Overall, "Option A" & "Option A2" are worrying as they encourage consumers to go against v9 philosophy around bundle size and remind me themes in v0:
- Current: "pay as you go", more components => more bundle size
- Proposed: "pay as you go" + _"global overrides without tree-shaking"_, more components => more bundle size + a piece on the top of application that can't be removed

Reply:
So an implementation that is a single method would have a switch but would end up depending on each of the components it customizes. This would prevent tree shaking out those components, right?

The property per component idea you have below that defined separate methods would avoid this singular dependency and allow tree-shaking right?

Reply:
> So an implementation that is a single method would have a switch but would end up depending on each of the components it customizes. This would prevent tree shaking out those components, right?
> The property per component idea you have below that defined separate methods would avoid this singular dependency and allow tree-shaking right?

I am referring to the following scenario (https://codesandbox.io/s/determined-spence-fufbef):
- We have an app that loads some components lazily
- _The main part_ of an app uses `FluentProvider` & `Button` (light components)
- _The lazy part_ of an app uses `Popover`, `Menu` & `Table` (heavy components)
- I want to use this option to customize components

I have customizations for `Popover` & `Button`. What can I do?
- Place `CustomizationProvider` on top of an app before `FluentProvider` ➡️ customizations for `Popover` will be loaded even when I don't need them and will be in _the main part_
- Place `CustomizationProvider` on top of an app with `Button` overrides & inside _the lazy part_ with `Popover` overrides

The first option will affect bundle size more & more if overrides will grow as they will be always loaded. The second option needs more maintenance (if `Popover` will appear in _the main part_, overrides should be moved). May be I overestimate the problem, but this option is really powerful.

Reply:
I updated OptionA2 to follow your suggested approach of separate hooks per component and passing state:unknown.